Creating a real WindowsService for MessagingServer

Stelzi79 opened the pull request:
MessagingServer.exe can install, uninstall, start and stop the WindowsService now
MessagingServer.exe needs and forces AdminRights

Stelzi79 committed:
The ServerPort wasn't saved between programm-starts

Stelzi79 committed:
initial works on the WindowsService

Stelzi79 committed:
Service works with installing it with on DeveloperComandLine with installutil

Stelzi79 committed:
MessagingServer.exe can install, uninstall, start and stop the WindowsService now
MessagingServer.exe needs and forces AdminRights

Stelzi79 committed:
modify setup.iss

emreeren commented:
Hello Wolfgang. 
1. I received some encoding errors from project files. It disappeared after I manually edited and saved project files. I think some German characters in project file comments caused this. I wish visual studio have a setting for using English language on auto generated files, otherwise nobody will be able to open these projects without solving these encoding problems.
2. I received a certificate error and it disappeared after I unchecked "Sign Clickonce Manifests" setting from project settings > signing.
3. It couldn't find Samba.messagingserver.windows service so I tried adding it as a reference but it generated another problem since windowsservice already references messaging server. I found it just references it for reading port setting. I've hardcoded port number and switched references for testing.
4. After doing these it installs fine but I'm getting "Access Denied" error when I try to start messaging server. I tried installing and running it as administrator but no change. I think I did something unexpected. 

Can't we keep both options side by side? I mean running it as a windows service is the right thing to do but in some cases running it as a standalone exe is a simpler solution for novice users.

Stelzi79 commented:
Hm some errors needed to be ironed out.
With the portnumber I will change.
For the access denied wich os do you have and have you activated UAC. My understanding was that the MassesingServer.exe force you now to start it as a Administrator - without elevated right it can't access all things right. I've only tested it so far on my Windows 8 with activated UAC (doh with out UAC no MetroApp starts). Operating a POS-Station with always administrive rights and without UAC and Firewall (at least the Windows one) is a very dangerous road to travel on. The big portion of the bussines depends on a functiniing POS-System and should be handled as mission-critical

I think the WindowsService-Way is the wright and the only way to do it espacialy for novices! I have now the experience that even have to manualy close the Workingperiod is for some waiters a hurdle and they ask for a automatic start and end of the workingperiod. I already had the conversation: "Your new software is _beep_ _beep_ _beep_ and has brocken the network as it says in the error on the bottom!" To explain to such people why this error occours is nearly impossible. As a supporting standpoint it's better when I sett it up once and on every reboot of the system it starts the service automaticaly without having to even loggin as it should be on a ServerOS.

emreeren commented:
I'm on windows 8, UAC disabled and my account is administrator.

I agree if installed properly service solution is better but I still want to keep standalone solution on hand.

Stelzi79 committed:
- configure now the port in a file in the AppData-Directory
- moved ServiceHelper to the WindowsService
- fixed a FileNotFoundException on uninstall of the Service
- removed the automatic start of the manual(old) service if it was setted before
- the MessagingServer.exe is now only a tool for servicing the WindowsService

Stelzi79 commented:
I've updated this branch and you were wright about the encoding errors. Somehow my process of take the patch which created hg and patch with it the git-repository has an encoding issue

I also fixed some other issues. I also completly removed any functions of the old one because some functions (the autostarting thingy in particular) I think did go haywire on your system. (I will include the old standalone-functons later again)

Be aware that you can't debug and step through a WindowsService or let it view an UI-Element!

I've tested it today on a Windows 8 Pro x86, Windows 8 Enterprise x64 and Windows Server 2012 Datacenter x64 (all with standard out of the box UAC-settings) and except the autostarting thingy (not counting adding a Firewall-Rule) all worked as expected.

I've compiled a setup.exe with the newest additions https://bitbucket.org/Stelzi79/sambapos-3_stelzi79/downloads/2013-07-19%200110-SambaSetup3015-StW_Beta.exe and see if the sourcecode compiles on your system without any errors.

Stelzi79 commented:
By the way if all other developers cared about accessing only the wright places UAC wouldn't be such a hassle to deal with. For example writing to the "Program Files"-Directory all the time wihtout a real cause and not to ProgramData needs an admin and triggers the UAC.

emreeren commented:
Your setup works fine for me however when I try to build from your branch it generates same "access denied" error. It also looks a little different on design time.

![2013-07-19 02_57_52-overlaycontrol](https://f.cloud.github.com/assets/835381/823223/ef6d8a7a-f005-11e2-88bd-f77b600c45c8.jpg)

I have to disable "Sign Clickonce Manifests" setting to be able to build it. Should I enable it?

I'm planning to release 3.0.15 today but I can't release it without the standalone server because it may break some users testing it in production. So I want to postpone this commit to the next release.

Stelzi79 commented:
I start my Visual Studio as an Admin so I don't know if that is the cause. Can you check out the source code from my hg-repository at https://bitbucket.org/Stelzi79/sambapos-3_stelzi79 (the StW_MessagingServerWinService branch) and see if you have the same issues.

I also test it with release build settings.

It isn't ready so it shouldn't be in a release anyway.

Stelzi79 committed:
- removed ClickOnce and regenerated some files in the english-version of VisualStudio

Stelzi79 committed:
Renamed MessagingServer to MessagingServerServiceTool

Stelzi79 committed:
readded the original MessagingServer

Stelzi79 committed:
changes to setup.iss

emreeren commented:
Wow great improvement! Thanks.
I found the reason of the access denied problem. I've changed ProcessInstaller account from LocalService to LocalSystem so it installed and started fine. Great!

Finally what I'm wondering is the updates. How can we install new sambapos version while a service is already running on the server? I think we should stop (and uninstall?) services with inno setup. Maybe we can also install services through inno setup since it already have admin rights it will be easier.

emreeren commented:
I think people uses this script for managing services from inno setup.
http://www.sandon.it/?q=node/9

Stelzi79 commented:
Yes I think the inno setup has to uninstall an already older installed service but shouldn't install such a service by default, because the 8080-port is a quiet common port used for example by every proxy-service like firewall-passthroughts.Therefore I also think we should change this default port to another more "exotic" port.

I also think we should include a deployment/update-solution to SambaPOS. I've a few idias to manage several POS-Satations and update them all with one centrialized click on a update-button. To access one DB from several different Client-Version is a very scary thing for me. Such a deployment/update-solution could also handle the WindowsService.

From a suporting-standpoint it is very tedious to have to physicaly run the setup on every POS-Stattion and normal User (not us) only installs and updates only flashy maleware and a legitimate software can only do a automatic unasked Update to be updated. Messages and manual updates from legitimate software even if they plug big security-holes are beeing ignored till the hardware refuses to work.

emreeren commented:
OK I've merged and updated some encoding issues. Thanks.
